tools/codegen/core/gen_huffman_decompressor.cc

```diff
@@ -1660,46 +1690,45 @@ int main(void) {
 
-  auto index_bzl = std::make_unique<Sink>();
-  index_bzl->Add<Prelude>("#", 2023);
-  index_bzl->Add(
-      "load(\"//bazel:grpc_build_system.bzl\", \"grpc_cc_library\", "
-      "\"grpc_package\")");
-  index_bzl->Add("licenses([\"notice\"])");
-  index_bzl->Add(
-      "grpc_package(name = \"test/cpp/microbenchmarks/huffman_geometries\", "
-      "visibility = \"public\")");
-
-  index_bzl->Add("grpc_cc_library(");
-  index_bzl->Add("  name = \"huffman_geometries\",");
-  index_bzl->Add("  srcs = [");
-  auto index_srcs = index_bzl->Add<Sink>();
-  index_bzl->Add("  ],");
-  index_bzl->Add("  hdrs = [");
-  index_bzl->Add("    \"index.h\",");
-  auto index_hdrs = index_bzl->Add<Sink>();
-  index_bzl->Add("  ],");
-  index_bzl->Add("  deps = [\"//:gpr_platform\"],");
-  index_bzl->Add(")");
-
   for (auto& r : results) {
-    index_includes->Add(absl::StrCat("#include \"", r->header_name, "\""));
-    index_decls->Add(absl::StrCat("  DECL_BENCHMARK(grpc_core::", r->ns,
-                                  "::HuffDecoder, ", r->ns, "); \\"));
-    index_hdrs->Add(
-        absl::StrCat("    \"", SplitAfter(r->header_name, '/'), "\","));
-    index_srcs->Add(
-        absl::StrCat("    \"", SplitAfter(r->source_name, '/'), "\","));
-    WriteFile(r->header_name, r->header);
-    WriteFile(r->source_name, r->source);
+    r->AddTailMatter();
+    index_includes->Add(absl::StrCat("#include \"", r->base_name, ".h\""));
+    for (const auto& ns : r->all_ns) {
+      index_decls->Add(absl::StrCat("  DECL_BENCHMARK(grpc_core::", ns,
+                                    "::HuffDecoder, ", ns, "); \\"));
+    }
+    WriteFile(r->base_name + ".h", r->header);
+    WriteFile(r->base_name + ".cc", r->source);
   }
   WriteFile("test/cpp/microbenchmarks/huffman_geometries/index.h",
             index_hdr->ToString());
-  WriteFile("test/cpp/microbenchmarks/huffman_geometries/BUILD",
-            index_bzl->ToString());
+}
+
+void GenSelected() {
+  FileSet selected("src/core/ext/transport/chttp2/transport/decode_huff");
+  selected.AddFrontMatter(2023);
+  selected.AddBuild(std::vector<int>({15, 7, 8}), true);
+  selected.AddTailMatter();
+  WriteFile(selected.base_name + ".h", selected.header);
+  WriteFile(selected.base_name + ".cc", selected.source);
+}
 
-  auto selected = Build(std::vector<int>({15, 7, 8}), true, 2022);
-  WriteFile(selected.header_name, selected.header);
-  WriteFile(selected.source_name, selected.source);
+int main(int argc, char** argv) {
+  bool did_stuff = false;
+  std::map<std::string, std::function<void()>> generators = {
+      {"microbenchmarks", GenMicrobenchmarks}, {"selected", GenSelected}};
+  for (int i = 1; i < argc; i++) {
+    auto it = generators.find(argv[i]);
+    if (it == generators.end()) {
+      fprintf(stderr, "Unknown generator: %s\n", argv[i]);
+      return 1;
+    }
+    it->second();
+    did_stuff = true;
+  }
+  if (!did_stuff) {
```

Comment:
if `argc > 1`, then either `did_stuff = true` or it returned early. You can just check argc.

Reply:
nice, done